cached_addr_book.go

```diff
@@ -0,0 +1,307 @@
+	cab.isProbing.Store(true)
+	defer cab.isProbing.Store(false)
+
+	start := time.Now()
+	defer func() {
+		duration := time.Since(start).Seconds()
+		probeDurationHistogram.Observe(duration)
+		logger.Debugf("Finished probing peers in %s", duration)
+	}()
+
+	var wg sync.WaitGroup
+	// semaphore channel to limit the number of concurrent probes
+	semaphore := make(chan struct{}, MaxConcurrentProbes)
+
+	for i, p := range cab.addrBook.PeersWithAddrs() {
+		if hasValidConnectedness(host.Network().Connectedness(p)) {
+			continue // don't probe connected peers
+		}
+
+		if !cab.ShouldProbePeer(p) {
+			continue
+		}
+
+		addrs := cab.addrBook.Addrs(p)
+
+		if !cab.allowPrivateIPs {
+			addrs = ma.FilterAddrs(addrs, manet.IsPublicAddr)
+		}
+
+		if len(addrs) == 0 {
+			continue // no addresses to probe
+		}
+
+		wg.Add(1)
+		go func() {
+			semaphore <- struct{}{}
+			defer func() {
+				<-semaphore // Release semaphore
+				wg.Done()
+			}()
+			probedPeersCounter.Inc()
+			ctx, cancel := context.WithTimeout(ctx, ConnectTimeout)
+			defer cancel()
+			logger.Debugf("Probe %d: PeerID: %s, Addrs: %v", i+1, p, addrs)
+			// if connect succeeds and identify runs, the background loop will take care of updating the peer state and cache
+			err := host.Connect(ctx, peer.AddrInfo{
+				ID:    p,
+				Addrs: addrs,
+			})
+			if err != nil {
+				logger.Debugf("failed to connect to peer %s: %v", p, err)
+				cab.RecordFailedConnection(p)
+			}
+		}()
+	}
+	wg.Wait()
+}
+
+// Returns the cached addresses for a peer, incrementing the return count
+func (cab *cachedAddrBook) GetCachedAddrs(p peer.ID) []types.Multiaddr {
+	cachedAddrs := cab.addrBook.Addrs(p)
+
+	if len(cachedAddrs) == 0 {
+		return nil
+	}
+
+	var result []types.Multiaddr // convert to local Multiaddr type 🙃
+	for _, addr := range cachedAddrs {
+		result = append(result, types.Multiaddr{Multiaddr: addr})
+	}
+	return result
+}
+
+// Update the peer cache with information about a failed connection
+// This should be called when a connection attempt to a peer fails
+func (cab *cachedAddrBook) RecordFailedConnection(p peer.ID) {
+	pState, exists := cab.peerCache.Get(p)
+	if !exists {
+		pState = peerState{}
+	}
+	pState.lastFailedConnTime = time.Now()
+	pState.connectFailures++
+	cab.peerCache.Add(p, pState)
+}
+
+// Returns true if we should probe a peer (either by dialing known addresses or by dispatching a FindPeer)
+// based on the last failed connection time and connection failures
+func (cab *cachedAddrBook) ShouldProbePeer(p peer.ID) bool {
+	pState, exists := cab.peerCache.Get(p)
+	if !exists {
+		return true // default to probing if the peer is not in the cache
+	}
+
+	var backoffDuration time.Duration
+	if pState.connectFailures > 0 {
+		// Calculate backoff only if we have failures
+		// this is effectively 2^(connectFailures - 1) * PeerProbeThreshold
+		// A single failure results in a 1 hour backoff
+		backoffDuration = PeerProbeThreshold * time.Duration(1<<(pState.connectFailures-1))
```

Comment:
NIT: Should this have an upper bound?

Reply:
Maybe. eventually it'll either get cleaned from the cache if it reaches max capacity or will not even reach here if 48 hours pass (and the peer TTL expires in the address book) unless the peer is encountered again via a FindPeer/FindProviders call to the service.

Reply:
I'd set the BackoffDuration to `min(backoffDuration, MaxBackoffDuration)`. This would mean eventually we will probe a peer once a day(at least). The current method is also fine though. I don't have a strong opinion here.